packages/ingest/recursion.js

```diff
@@ -1,84 +1,93 @@
-      // is a regex rule
-      const isRegex = map[position + 1];
+  const path = currentPath.replace(/\/+/g, '/');
+  const contents = fn(path);
+  let files = [];
 
-      // if it is the regex rule, we use the rule to
-      // decide whether to do more recursion
-      if (isRegex) {
-        regexPath = new RegExp(rules[position + 1]);
-      } else {
-        // if it is just a path, we use that segment and its children
-        textPath = rules[position + 1];
+  for (let ctr = 0; ctr < contents.length; ctr += 1) {
+    const item = contents[ctr];
+    if (['-', 0].includes(item.type)) {
+      if (!isRegex || item.name.match(new RegExp(filterExpr))) {
+        // add file to the list if it matches rule
+        files.push(item);
       }
-
-      //  if next segment is regex and matches the rule
-      //  list
-      if (isRegex && item.name.match(regexPath)) {
-        newPath = join(path, item.name);
-      } else if (textPath) {
-        // if it is a regular path, use the path
-        newPath = join(path, textPath);
-      } else {
-        // and ignore all cases that does't match this rule
-        continue; // eslint-disable-line no-continue
+    } else if (['d', 1].includes(item.type)) {
+      // check if dir matches rule
+      if (!isRegex || item.name.match(new RegExp(filterExpr))) {
+        // eslint-disable-next-line no-await-in-loop
+        files = files.concat(await recurOnDirectory(fn, `${path}/${item.name}/`, segments, position + 1));
       }
-
-      // eslint-disable-next-line no-await-in-loop
-      const tmp = await recursion(fn, originalPath, newPath, position + 1);
-      files = files.concat(tmp);
-
-      if (textPath) break;
-    } else if (item.type === '-' || item.type === 0) {
-      // add file to the list
-      files = files.concat(item);
     }
   }
-
   return files;
 }
 
+/**
+ * Recursively list contents of a directory, filtering on provided regex segments.
+ *
+ * @param {Function} fn - list function
+ * @param {string} path - path string which may contain regexes for filtering
+ */
+async function dynamicRecursiveList(fn, path) {
+  const dynamicRegex = /(\([^\)]*\))/g;
+  const normalizedPath = normalizeWithSlashes(path);
+  const segments = normalizedPath
+    .split(dynamicRegex) // split into text path and regex segments
+    .filter((i) => !!i); // filter out empty segments from split
+
+  return recurOnDirectory(fn, segments[0], segments, 0);
+}
+
+/**
+ * Handles recursion of a FTP/SFTP list operation
+ * It requests a promisified list function that returns contents of
+ * a directory on a server
+ *
+ * @param {function} fn The promisified function for listing a directory
+ * @param {string} originalPath the full path to use for recursively listing the directory
+ * @returns {Promise} the promise of an object that has the path is the key and
+ *   list of files as values
+ */
+async function recursion(fn, originalPath) {
+  return dynamicRecursiveList(fn, originalPath);
```

Comment:
Why is this calling `dynamicRecursiveList` instead of just including the logic of `dynamicRecursiveList`? In other words, why the extra function nesting?

Reply:
Since `module.exports = recursion`, there is no way to change the exports without creating a breaking change. We can do that, if you feel strongly that we should.

Reply:
No, I don't think we need to do that

Reply:
I still don't follow. Why don't you simply put the contents of the function `dynamiceRecursiveList` into the function `recursion` and then delete `dynamicRecursiveList`? This doesn't change the exports, it just eliminates an unnecessary function, no?

Reply:
Because `recursion` is an awful function name. 😅I guess it's having my cake and eating it not to break the exports.

Reply:
Yeah, perhaps not the best name, but personally I'd prefer not to bother with creating an unnecessary function with a name that you like since you still have to keep the name you don't like. The unnecessary function just leads to unnecessary confusion and a line of unnecessary questioning, such as this.

However, if you really, really want to keep your unnecessary function, at least remove the duplicated jsdoc on your unnecessary function since it actually does not match the jsdoc on `recursion`, and this avoids having to keep the jsdocs in sync. And if you do keep the function, you could simplify a bit like so:

```
<remove jsdoc>
// <non-jsdoc note about why this is here when it doesn't really need to be>
async function dynamicRecursiveList(fn, path) {
  ...
}

/**
 * <complete (including @return) and correctly formatted jsdoc>
 */
const recursion = dynamicRecursiveList;

module.exports = recursion;
```

While you're at it, also add the missing `@return`s to the other functions.